Add support for parsing direct conversations

cerber-os opened the pull request:
### What's changed
- Added support for 1-1 private conversations (`-u` argument)

### Testing method
Created backup of random private conversation and verified its content with the data displayed on Wire Desktop app.

### Resolved
PR implements feature suggested in #2

cerber-os committed:
Add support for parsing direct conversations

cerber-os committed:
Update README.md

Lachcim commented on wirebackupparser/groups.py:
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
+    """
+    Despite what name might suggest, this class handles both direct messages and group conversations.
+    Don't blame me, this is how Wire developers decided to implement backup format
+    """
```
Wire refers to these entities as "conversations". There are 4 types of conversations: `GROUP`, `SELF`, `ONE2ONE` and `CONNECT`. I propose renaming every instance of "group" to "conversation", except where relevant.

cerber-os replied:
Good idea - commit 288dac4 renamed classes `Group(s)` to `Conversation(s)`

Lachcim commented on wirebackupparser/groups.py:
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
+                                         backupOwner=backup.backupOwner))
+            elif conv.get('type') == ENT_TYPE_DIRECT:
+                self.directGroups.append(Group(conv.get('id'),
+                                               name="Direct conversation",
```
The conversation name is relevant to the report and should not be neglected. A proper name can be obtained via `getUserByID(...).name`. Note: do not use `conv.get('name')`, those are not unique and depend on who started the conversation (i.e. sent a connection request).

cerber-os replied:
Implemented in 288dac4

Lachcim commented on wirebackupparser/groups.py:
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
+            elif conv.get('type') == ENT_TYPE_DIRECT:
+                self.directGroups.append(Group(conv.get('id'),
+                                               name="Direct conversation",
+                                               creator=None,
```
This information is not meaningless, see above

cerber-os replied:
The same as above

Lachcim commented on wirebackupparser/groups.py:
```diff
@@ -53,3 +65,12 @@ def getGroupByName(self, name):
+        for directGroup in self.directGroups:
+            if directGroup.members[0] == userID:
+                return directGroup
+        raise KeyError("Unknown direct group with member {}".format(userID))
```
I see this as an attempt to gracefully fail when a user lookup fails, but currently this only works for resolved usernames with whom a connection hasn't been established (and thus there isn't a one-to-one conversation).

cerber-os replied:
This exception is raised when user with provided nickname is a member of one of the group but not direct conversation.
In such cases, his nickname is properly translated to `userID` (\`cause he exists in `events.json`) but no direct conversation is present in `self.directGroup` list.

Lachcim replied:
Correct, that's what I meant

Lachcim reviewed (changes requested):
Minor issues with error handling and conversation metadata parsing. Additional code refactoring (rename "group" to "conversation") suggested.

Lachcim commented:
Two remarks regarding the current state of code:

* We are throwing exceptions in response to erroneous user input, but we are not catching them anywhere. A top-level try-except statement can be added to gracefully display error messages.
* We inform the user of ambiguous usernames, but we provide no method of resolving this ambiguity. I propose implementing additional command-line arguments (`-gid`, `-uid`?) for specifying group and user IDs explicitly. Also handles text encoding issues.

Lachcim committed:
Fix graphing bug when total is 0

cerber-os committed:
Fix issues with parsing direct conversations

- Rename Group(s) classes to Conversation(s)
- Fix error handling
- Store creator and name of direct conversations
- Add option -uid

cerber-os commented on wirebackupparser/events.py:
```diff
@@ -23,6 +23,13 @@ def getUserByID(uid):
+    result = [user for user in Users if user.name == name]
+    if len(result) > 1:
+        raise KeyError("Ambigious username {} - found {} times".format(name, len(result)))
+    return result[0]
```
Fixed by 288dac4

cerber-os commented:
> We are throwing exceptions in response to erroneous user input, but we are not catching them anywhere. A top-level try-except statement can be added to gracefully display error messages.

Good idea, but such feature is out of the scope of this pull request - consider creating a new issue (and possibly assign yourself).

> We inform the user of ambiguous usernames, but we provide no method of resolving this ambiguity. I propose implementing additional command-line arguments (`-gid`, `-uid`?) for specifying group and user IDs explicitly.

As far as I know, Wire prohibits ambiguous usernames by showing `Nickname is already taken` message on settings page. Therefore, I'm not sure if such options would be in any way useful in real applications.

The checks in Python code are a bit excessive, but should provide some leads when dealing with damaged/improper backup file.

> Also handles text encoding issues.

Could you elaborate?

Lachcim commented:
I was able to trigger a `KeyError: 'Ambigious username 3na10 - found 2 times'` by renaming an account in the preferences screen. If you believed this to be impossible, why check against it with an if statement?
![Annotation 2020-08-03 205109](https://user-images.githubusercontent.com/15709471/89216801-5cc96000-d5cb-11ea-9dd5-0b05e043ec06.png)

Encoding issues appear when you try to type non-ASCII characters in various terminal environments. For example, on Windows, the command `echo zażółć gęślą jaźń > test.txt` outputs `zaz¢lc gesla jazn` when the codepage is set to 437. This may prohibit some users from querying certain user and group names.

cerber-os commented:
Oh, you're right - Wire disallows duplicated nicknames (i.e. those prefixed with `@`), but seems to work fine with not ambiguous usernames (thanks, Wire dev team!). I'll try to add some kind of selection when more than one matching user was found.

As for the encoding issues: if your terminal doesn't support Unicode after almost 30 years since the first release of ISO10646 standard, I kindly suggest to change it...

cerber-os commented:
Now, when non-unique username is provided the following message will be displayed:
```txt
./parser.py -f wire.backup -u Smith
Loading backup...
Parsing backup...
Available IDs for username Smith:
	- 4c676a26-e5e9-44ae-a738-61694cc693d4
	- 43ae3b0a-6549-47b3-9e9d-43978f54fa9d
'Ambiguous username Smith - found 2 times'
[+] Try again with option -uid
```

Newly added option `--uid` allows to select one of the available IDs:
```sh
./parser.py -f wire.backup --uid 43ae3b0a-6549-47b3-9e9d-43978f54fa9d
```

Lachcim committed:
Fix typo

Lachcim reviewed (approved):
Excellent work as expected of Mr eng. Wieczorek himself. Thank you so much for this exceptional contribution. You are a scholar and a gentleman.

h0adp0re commented:
Fantastic, thank you!
